File: 프로그래머스/2/181187. 두 원 사이의 정수 쌍/두 원 사이의 정수 쌍.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <math.h>

long long solution(int r1, int r2) {
    long long answer = 0;

    long long R1 = (long long)r1;
    long long R2 = (long long)r2;
    long long R1sq = R1 * R1;
    long long R2sq = R2 * R2;

    for (int x = 0; x <= r2; x++) {
        long long xx = (long long)x * x;

        // 바깥 원에서의 최대 y
        long long maxY = (long long)floor(sqrt((double)(R2sq - xx)));

        // 안쪽 원에서의 최소 y
        long long minY;
        long long diff1 = R1sq - xx;

        if (diff1 > 0) {
            minY = (long long)ceil(sqrt((double)diff1));
        } else {
            minY = 0;
        }

        if (maxY < minY) continue;  // 해당 x에서 유효한 y 없음

        if (x == 0) {
            // y축 위: (0, y), (0, -y) → 각 y당 2개
            answer += 2 * (maxY - minY + 1);
        } else {
            if (minY == 0) {
                // y = 0: (±x, 0) → 2개
                answer += 2;
                // y >= 1: (±x, ±y) → y 하나당 4개
                answer += 4 * maxY;
            } else {
                // y >= 1: (±x, ±y) → y 하나당 4개
                answer += 4 * (maxY - minY + 1);
            }
        }
    }

    return answer;
}

File: 프로그래머스/2/181187. 두 원 사이의 정수 쌍/두 원 사이의 정수 쌍.java
class Solution {
    public long solution(int r1, int r2) {
        long answer = 0;

        long R1 = (long) r1;
        long R2 = (long) r2;
        long R1sq = R1 * R1;
        long R2sq = R2 * R2;

        for (int x = 0; x <= r2; x++) {
            long xx = (long) x * x;

            // 바깥 원에서의 최대 y
            long maxY = (long) Math.floor(Math.sqrt(R2sq - xx));

            // 안쪽 원에서의 최소 y
            long minY;
            long diff1 = R1sq - xx;
            if (diff1 > 0) {
                minY = (long) Math.ceil(Math.sqrt(diff1));
            } else {
                minY = 0;
            }

            if (maxY < minY) continue; // 이 x에서는 유효한 점 없음

            if (x == 0) {
                // y축 위: (0, y), (0, -y) → 각 y마다 2개
                answer += 2L * (maxY - minY + 1);
            } else {
                if (minY == 0) {
                    // y = 0: (±x, 0) → 2개
                    answer += 2L;
                    // y >= 1: (±x, ±y) → y 하나당 4개
                    answer += 4L * maxY;
                } else {
                    // y >= minY: (±x, ±y) → y 하나당 4개
                    answer += 4L * (maxY - minY + 1);
                }
            }
        }

        return answer;
    }
}